flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -106,14 +106,24 @@ static bool isAbsent(const fir::ExtendedValue &exv) {
   return !fir::getBase(exv);
 }
 
-/// Process calls to Maxval, Minval intrinsic functions
+/// Placeholder function for SUM/PRODUCT intrinsics when genSum()
+/// or genProduct() call genExtremumValProdSum(). It is an internal
+/// compiler error (ICE) if this function gets called.
+static void genInvalidChar(Fortran::lower::FirOpBuilder &builder,
+                           mlir::Location loc, mlir::Value resultBox,
+                           mlir::Value arrayBox, mlir::Value maskBox,
+                           llvm::StringRef errMsg) {
+  fir::emitFatalError(loc, errMsg);
```

Comment:
Ah. This is a consequence of trying to merge MAX/MIN and PROD/SUM, right? It seems like it might be better to leave these separated rather than merge the lowering code gen.